File: C++/Grind75/Week-4/208.implement-trie-prefix-tree.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=208 lang=cpp
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
struct Node
{
    Node *links[26];
    bool flag = false;
    bool containsKey(char ch)
    {
        return links[ch - 'a'] != NULL;
    }

    void insertKey(char ch, Node *node)
    {
        links[ch - 'a'] = node;
    }

    Node *getNext(char ch)
    {
        return links[ch - 'a'];
    }

    void setEnd()
    {
        flag = true;
    }

    bool getEnd()
    {
        return flag;
    }
};

class Trie
{
private:
    Node *root;

public:
    Trie()
    {
        root = new Node();
    }

    void insert(string word)
    {
        Node *node = root;
        for (int i = 0; i < word.size(); i++)
        {
            if (!(node->containsKey(word[i])))
            {
                node->insertKey(word[i], new Node());
            }
            node = node->getNext(word[i]);
        }
        node->setEnd();
    }

    bool search(string word)
    {
        Node *node = root;
        for (int i = 0; i < word.size(); i++)
        {
            if (!(node->containsKey(word[i])))
            {
                return false;
            }
            node = node->getNext(word[i]);
        }
        return node->getEnd();
    }

    bool startsWith(string prefix)
    {
        Node *node = root;
        for (int i = 0; i < prefix.size(); i++)
        {
            if (!(node->containsKey(prefix[i])))
            {
                return false;
            }
            node = node->getNext(prefix[i]);
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
// @lc code=end

File: Java/Grind75/Week 4/208.implement-trie-prefix-tree.java
import java.util.*;

/*
 * @lc app=leetcode id=208 lang=java
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
class Trie {

    Trie[] trie;
    Boolean flag;

    public Trie() {
        trie = new Trie[26];
        flag = false;
    }

    public void insert(String word) {
        if (word.length() == 0) {
            flag = true;
            return;
        }
        char ch = word.charAt(0);
        if (trie[ch - 'a'] == null) {
            trie[ch - 'a'] = new Trie();
        }
        trie[ch - 'a'].insert(word.substring(1));
    }

    public boolean search(String word) {
        if (word.length() == 0) {
            return flag;
        }
        char ch = word.charAt(0);
        if (trie[ch - 'a'] == null) {
            return false;
        }
        return trie[ch - 'a'].search(word.substring(1));
    }

    public boolean startsWith(String prefix) {
        if (prefix.length() == 0) {
            return true;
        }
        char ch = prefix.charAt(0);
        if (trie[ch - 'a'] == null) {
            return false;
        }
        return trie[ch - 'a'].startsWith(prefix.substring(1));
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
// @lc code=end
